control-plane-operator/hostedclusterconfigoperator/controllers/globalps/globalps.go

```
@@ -113,37 +165,128 @@ func (r *Reconciler) reconcileGlobalPullSecret(ctx context.Context) error {
 	daemonSet := manifests.GlobalPullSecretDaemonSet()
-	if err := reconcileDaemonSet(ctx, daemonSet, secret.Name, r.hcUncachedClient, r.CreateOrUpdate, r.hccoImage); err != nil {
+	if err := reconcileDaemonSet(ctx, daemonSet, secret.Name, originalSecret.Name, configSeed, r.hcUncachedClient, r.CreateOrUpdate, r.hccoImage); err != nil {
 		return fmt.Errorf("failed to reconcile global pull secret daemon set: %w", err)
 	}
 
 	return nil
 }
 
-func reconcileDaemonSet(ctx context.Context, daemonSet *appsv1.DaemonSet, globalPullSecretName string, c crclient.Client, createOrUpdate upsert.CreateOrUpdateFN, hccoImage string) error {
+// labelNodesFromInPlaceNodePools labels nodes that belong to NodePools using InPlace upgrade strategy
+func (r *Reconciler) labelNodesFromInPlaceNodePools(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Get all nodes from the hosted cluster
+	nodeList := &corev1.NodeList{}
+	if err := r.hcUncachedClient.List(ctx, nodeList); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	// Get all MachineSets to identify which use InPlace upgrade strategy
+	machineSetList := &capiv1.MachineSetList{}
+	if err := r.cpClient.List(ctx, machineSetList, &crclient.ListOptions{
+		Namespace: r.hcpNamespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list MachineSets: %w", err)
+	}
+
+	// Create set of nodes that belong to InPlace NodePools
+	nodesFromInPlaceNodePools := make(map[string]bool)
+	for _, ms := range machineSetList.Items {
+		// Check if this MachineSet belongs to a NodePool with InPlace strategy
+		// This can be identified by the presence of InPlace-specific annotations
+		if _, hasTargetConfig := ms.Annotations["hypershift.openshift.io/nodePoolTargetConfigVersion"]; hasTargetConfig {
+			// Get Machines from this MachineSet
+			machines := &capiv1.MachineList{}
+			if err := r.cpClient.List(ctx, machines, &crclient.ListOptions{
+				Namespace:     ms.Namespace,
```

🔍 **Node Labeling Logic Fragility**

This function relies on annotation detection that could be fragile:

```go
if _, hasTargetConfig := ms.Annotations["hypershift.openshift.io/nodePoolTargetConfigVersion"]; hasTargetConfig {
    // Assumes this annotation identifies InPlace strategy
}
```

**Concerns**:
- Fragile dependency on annotation naming conventions
- No validation that this annotation correctly identifies InPlace strategy
- Silent failures when node labeling fails could lead to DaemonSet being deployed to wrong nodes

**Recommendation**:
- Add more robust NodePool strategy detection (check actual NodePool objects)
- Add validation and better error handling 
- Consider using explicit NodePool strategy identification rather than annotation inference